File: leetcode/1001-1999/1657-determine-if-two-strings-are-close/DetermineIfTwoStringsAreClose.java
import java.util.Arrays;

public class DetermineIfTwoStringsAreClose {
  public static void main(String[] args) {
    System.out.println(new Solution().closeStrings("abc", "bca"));
  }
}

class Solution {
  public boolean closeStrings(String lhs, String rhs) {
    if (lhs.equals(rhs)) {
      return true;
    }

    if (lhs.length() != rhs.length()) {
      return true;
    }

    int[] left = new int[26];
    int[] right = new int[26];

    for (int i = 0; i < lhs.length(); i++) {
      left[lhs.charAt(i) - 'a']++;
    }

    for (int i = 0; i < rhs.length(); i++) {
      right[rhs.charAt(i) - 'a']++;
    }

    for (int i = 0; i < 26; i++) {
      if ((left[i] > 0 && right[i] == 0) || (right[i] > 0 && left[i] == 0)) {
        return false;
      }
    }

    Arrays.sort(left);
    Arrays.sort(right);

    return Arrays.equals(left, right);
  }
}

File: leetcode/1001-2000/1601-1700/1657-determine-if-two-strings-are-close/DetermineIfTwoStringsAreClose.ts
function closeStrings(lhs: string, rhs: string): boolean {
  if (lhs.length !== rhs.length) {
    return false;
  }

  if (lhs === rhs) {
    return true;
  }

  const convert = (s: string) => s.split("").reduce(
    (accumulated, currentValue) => {
      accumulated[currentValue] = (accumulated[currentValue] ?? 0) + 1;

      return accumulated;
    }, {}
  )

  const left = convert(lhs);
  const right = convert(rhs);

  console.log(Object.keys(left).sort().toString());
  console.log(Object.keys(right).sort().toString());

  return Object.keys(left).sort().toString() === Object.keys(right).sort().toString()
    && Object.values(left).sort().toString() === Object.values(right).sort().toString();
};

console.log(closeStrings("abc", "bca"));
